### ISSUE TITLE
Stop FFMPEG upscale video

### ISSUE THREAD

**Issue by jessica88888** [binoculars/aws-lambda-ffmpeg#36]
Dear Binoculars,

I'm trying to scale the video down into multiple resolutions while keeping the aspect ratio (Shrink to fit). If the original video resolution is smaller than the resolution specified in the command, skip transcoding and stop creating the file for that particular resolution.

For example:

I have a video with resolution 1920x1080

Output:
3840x2160 (Skip/Stop)
2560x1440 (Skip/Stop)
1920x1080 (Proceed)
1280x720 (Proceed)

I've tried many commands, but none of them really works. May I know is there any command that are able to achieve this?

Thank you.

Best regards,
Jessica

**Comment by binoculars** [comment]
I don't the answer. You might want to use the FFprobe results and do something with those in JavaScript. If you find an FFmpeg command that will do the same thing, please let me know.

**Comment by jessica88888** [comment]
Dear Binoculars,

I've got an answer from another place. 

> This can be done by making use of the crop filter, which will fail if if the cropped output dimensions are larger than the input. If the command fails, the output file will be created but its size will be 0.
> 
> Each variant will need to be converted using a separate command. A single command can't be used as if one of the pipeline fails, the entire command is aborted.
> 
> The crop filter should be used like this:

`ffmpeg -i sample.mp4 -c:a copy -vf crop=iw:'max(ih,2160)',scale=-2:2160 -movflags +faststart out2160.mp4 `

> The crop width is the same as the input width. The crop height is set to the maximum of the input height and the target height. If the input height is less than the target, the crop filter initialization fails and the command is aborted. If the input height is equal or greater, then the full width & height is passed through to the scale filter.

After modifying, I've tried locally with FFMPEG:

`ffmpeg -i sample.mp4 -vf thumbnail -vf scale='min(500\,iw):-2' -vframes 1 thumb500.png -vf scale='min(270\,iw):-2' -vframes 1 thumb270.png -vf crop=iw:'max(360\,ih)',scale=-2:360 -movflags +faststart out360.mp4 && ffmpeg -i sample.mp4 -vf crop=iw:'max(720\,ih)',scale=-2:720 -movflags +faststart out720.mp4 && ffmpeg -i sample.mp4 -vf crop=iw:'max(1080\,ih)',scale=-2:1080 -movflags +faststart out1080.mp4 && ffmpeg -i sample.mp4 -vf crop=iw:'max(1440\,ih)',scale=-2:1440 -movflags +faststart out1440.mp4 && ffmpeg -i sample.mp4 -vf crop=iw:'max(2160\,ih)',scale=-2:2160 -movflags +faststart out2160.mp4 `

For example, if I provide a 1920x1080 video file, it will output:

Thumbnail 270px (File created normally)
Thumbnail 500px (File created normally)
Video 360P (File created normally)
Video 720P (File created normally)
Video 1080P (File created normally)
Video 1440P (File created with **0B** file size)
Video 2160P  (**No file**)

However, when I try to modify the **FFMPEG_ARGS** Enviroment Variable at Lambda, I'm unable to get the command  to work with the function. 

I understand there're mistakes in the command, could you please guide me on this?

Thank you.

Best regards,
Jessica

**Comment by binoculars** [comment]
Currently, I've only got the Lambda function to run one FFmpeg command by default, but it takes multiple arguments. So, you won't be able to run all of those commands sequentially. I may be able to modify the function to run multiple FFmpeg commands, each with different arguments. However, if we can find a solution where running a single command with stream copies, that would be ideal. I'll look into this more as well.

**Comment by jessica88888** [comment]
Hi Binoculars,

Thanks for your reply. May I know what command I can use to output multiple file by modifying the default environment variable?

`-c:a copy -vf scale='min(320\,iw):-2' -movflags +faststart -metadata description=http://my.site/$KEY_PREFIX.mp4 out.mp4 -vf thumbnail -vf scale='min(320\,iw):-2' -vframes 1 out.png`

For example: 
Thumbnail 270px
Thumbnail 500px 
Video 360P
Video 720P 
Video 1080P
Video 1440P
Video 2160P

Thank you.

Best regards,
Jessica

**Comment by binoculars** [comment]
If you run this command, it will give you back images and videos scaled to 320px and 640px width.

```bash
ffmpeg -i sample.mp4 \
  -c:a copy -vf scale='min(320\,iw):-2' -movflags +faststart out_320.mp4 \ 
            -vf thumbnail -vf scale='min(320\,iw):-2' -vframes 1 out_320.png \
  -c:a copy -vf scale='min(640\,iw):-2' -movflags +faststart out_640.mp4 \
            -vf thumbnail -vf scale='min(640\,iw):-2' -vframes 1 out_640.png
```

You should be able to adapt it for your needs.

**Comment by jessica88888** [comment]
```
ffmpeg -i sample.mp4 \
  -c:a copy -vf scale='min(320\,iw):-2' -movflags +faststart out_320.mp4 \ 
            -vf thumbnail -vf scale='min(320\,iw):-2' -vframes 1 out_320.png \
  -c:a copy -vf scale='min(640\,iw):-2' -movflags +faststart out_640.mp4 \
            -vf thumbnail -vf scale='min(640\,iw):-2' -vframes 1 out_640.png
```
I've just tried the command on AWS, but it only output 2 files to the bucket, the 640p didn't appear in S3 bucket.

Original file: samplefile.mp4

Output:
samplefile.mp4 (320p)
samplefile.png (320px)

Would it be due to same output name and it's replaced? If yes, how can we specify the file name in the command? Thanks Binoculars.

**Comment by binoculars** [comment]
Right, I was providing that as an example to run locally. The `uploadFile` function, here: https://github.com/binoculars/aws-lambda-ffmpeg/blob/develop/platform/common.js#L202 is going to need to be refactored to support uploading multiple files of the same type.

**Comment by jessica88888** [comment]
I see, may I know is there any possibility to release an update for this issue? Thanks Binoculars.

**Comment by binoculars** [comment]
Yes, I can work on it. It might be a few days or more before it's ready to go. If you would like to submit a pull request, I'd be happy to work with you to get it merged in.

**Comment by jessica88888** [comment]
Sure, I can submit a pull request, but I'm afraid that I wouldn't be able to help much as I've very limited knowledge on all of these, hope you don't mind about that.

**Comment by jessica88888** [comment]
I've just submitted a pull request, I'm new to Github. Don't know if I've submitted the pull request in a correct way? 

https://github.com/binoculars/aws-lambda-ffmpeg/pull/37

Thanks Binoculars.

**Comment by binoculars** [comment]
So you want to make a fork, clone it locally, make a feature branch, make your changes, commit those, push that branch up to your fork and then submit a pull request against my repository with the changes from your branch.

This might help explain: https://egghead.io/courses/how-to-contribute-to-an-open-source-project-on-github

**Comment by jessica88888** [comment]
Dear Binoculars,

May I know are you able to release an update for this issue if I didn't submit a pull request? I just noticed that the PR has been closed, I think I've done something wrongly. Sorry for any inconvenience caused, I'm a newbie that are totally unfamiliar with all of these. Hope you'll understand.

Thank you.

Best regards,
Jessica

**Comment by binoculars** [comment]
That's okay. You have to make a new PR from your fork with the code that you wrote so that I can merge it into my repository.

**Comment by jessica88888** [comment]
Sorry again. The problem is I don't know how to modify the current code to solve this issue, but I have a rough idea. Maybe we can get the resolution info from the FFMPEG_ARGS / FFPROBE and append it to `${keyPrefix}.${extension}` accordingly? Thank you.

**Comment by binoculars** [comment]
Yes, I think you're on the right track. We need to think about how that can be generalized so that it's configurable and will work with the majority of use cases.

**Comment by jessica88888** [comment]
I think if we follow the resolution info from FFBROBE, it's less configurable by the user. Maybe specifying a special filename in FFMPEG_ARGS will triggers the output with custom name? For example:

`ffmpeg -i sample.mp4 \
  -c:a copy -vf scale='min(320\,iw):-2' -movflags +faststart appendthis_mycustomname320.mp4 \ 
            -vf thumbnail -vf scale='min(320\,iw):-2' -vframes 1 appendthis_mycustomname320.png \
  -c:a copy -vf scale='min(640\,iw):-2' -movflags +faststart appendthis_anyname640.mp4 \
            -vf thumbnail -vf scale='min(640\,iw):-2' -vframes 1 appendthis_anyname640.png`

**sample.mp4** is the original filename

If FFMPEG_ARGS comes with **appendthis_** then extract the string between **appendthis_** and **extension**, append to the output filename, so the output filename will be like:

sample_mycustomname320.mp4
sample_mycustomname320.png
sample_anyname640.mp4
sample_anyname640.png

**Comment by jessica88888** [comment]
Hi Binoculars, 

May I know what's your suggestion to overcome the issue of Lambda's very limited execution time (Max.: 5 minutes) ?

Thank you.

Best regards,
Jessica

**Comment by binoculars** [comment]
If you need more than 5 minutes, I'd suggest taking a look at AWS Elastic Transcoder or using FFmpeg with ECS.

**Comment by jessica88888** [comment]
May I know is this library able to work with ECS? If yes, may I know any documentation on this? Thanks Binoculars.

**Comment by binoculars** [comment]
I'm sure it could be adapted, but ECS is out of scope for this project. Check out https://aws.amazon.com/ecs/getting-started/ for more info.

**Comment by jessica88888** [comment]
Alright, thank you binoculars for being so helpful.

Best regards,
Jessica

**Comment by jessica88888** [comment]
Dear Binoculars,

I was googling all around to find the solution to overcome 5 minutes limitation.

I just came across this blog post, do you think this can be applied with this repo.?

If we can extend the lambda's time, then this repo. will be very powerful.

http://tech.marketenterprise.co.jp/2016/10/02/go-beyond-5-minutes-with-aws-lambda-using-python/

Thank you. 

Best regards,
Jessica

**Comment by binoculars** [comment]
There are a wide variety of use cases and a very narrow set of circumstances where that's feasible for this project. The main limitation is that you're going to run out of space in the temp directory before you even hit the 5 minute limit. If we had more space in the temp directory, it would be a different story, but then we run into another issue, which is that video manipulation is mostly linear and requires the entire file (unless it's been chunked ahead of time, which would likely require a non-lambda solution).

I think you're going to get a lot more mileage out of ECS or Elastic Transcoder if you're doing a lot of long/large videos.

**Comment by binoculars** [comment]
You could do something with AWS step functions, such as

1. Set up an S3 event to trigger the Step Function
1. Verify the video is valid (optional)
1. Execute several lambda functions in parallel with different filters/scaling commands (this could be all streaming in from S3 and out to S3 without ever touching the disk)
1. Do something when all parallel tasks are complete (notification, database update, etc)

You still have the problem of the 5 minute limitation, but you may be able to get more done in those 5 minutes if you parallelize it, and then you wouldn't need to work about the temp disk space problem as much.

Unfortunately, this scenario is an order of magnitude more complex, so it's not something I'm pursuing at this point, but perhaps in the future. Maybe you could learn and then teach me 😄

**Comment by jessica88888** [comment]
Thanks Binoculars, I will try to find out other solution to solve the limitation problem. Thank you for your kind support.

Best regards,
Jessica
